File: 45-jump-game-ii/jump-game-ii.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n=len(nums)
        jump=0
        near=0
        far=0
        while far<n-1:
            max_reach=0
            for i in range(near,far+1):
                max_reach=max(max_reach,i+nums[i])
            near=far+1
            far=max_reach
            jump+=1
        return jump
        

        


        

File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    vector<int> dp;
    int solve(vector<int>& nums,int i){
        int n=nums.size();
        if(i==n-1) return 0;
        if(i>=n) return INT_MAX;
        if(dp[i]!=-1) return dp[i];
        int ans=INT_MAX;
        for(int j=1;j<=nums[i];j++){
            int a=solve(nums,i+j);
            if(a!=INT_MAX)
            ans=min(ans,a+1);
        }
        return dp[i]=ans;
    }
    int jump(vector<int>& nums) {
        int n=nums.size();
        dp.assign(n,-1);
        return solve(nums,0);
    }
};